tests/gradients/test_parameter_shift.py

```diff
@@ -914,3 +914,286 @@ def cost_fn(x):
+    """Test that tapes ending with expval(H) can be
+    differentiated"""
+
+    def test_not_expval_error(self):
+        """Test that if the variance of the Hamiltonian is requested,
+        an error is raised"""
+        dev = qml.device("default.qubit", wires=2)
+
+        obs = [qml.PauliZ(0), qml.PauliZ(0) @ qml.PauliX(1), qml.PauliY(0)]
+        coeffs = np.array([0.1, 0.2, 0.3])
+        H = qml.Hamiltonian(coeffs, obs)
+
+        weights = np.array([0.4, 0.5])
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(weights[0], wires=0)
+            qml.RY(weights[1], wires=1)
+            qml.CNOT(wires=[0, 1])
+            qml.var(H)
+
+        tape.trainable_params = {2, 3, 4}
+
+        with pytest.raises(ValueError, match="for expectations, not var"):
+            qml.gradients.param_shift(tape)
+
+    def test_no_trainable_coeffs(self, mocker, tol):
+        """Test no trainable Hamiltonian coefficients"""
+        dev = qml.device("default.qubit", wires=2)
+        spy = mocker.spy(qml.gradients, "hamiltonian_grad")
+
+        obs = [qml.PauliZ(0), qml.PauliZ(0) @ qml.PauliX(1), qml.PauliY(0)]
+        coeffs = np.array([0.1, 0.2, 0.3])
+        H = qml.Hamiltonian(coeffs, obs)
+
+        weights = np.array([0.4, 0.5])
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(weights[0], wires=0)
+            qml.RY(weights[1], wires=1)
+            qml.CNOT(wires=[0, 1])
+            qml.expval(H)
+
+        a, b, c = coeffs
+        x, y = weights
+        tape.trainable_params = {0, 1}
+
+        res = dev.batch_execute([tape])
+        expected = -c * np.sin(x) * np.sin(y) + np.cos(x) * (a + b * np.sin(y))
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        tapes, fn = qml.gradients.param_shift(tape)
+        # two shifts per rotation gate, one circuit per trainable H term
+        assert len(tapes) == 2 * 2
+        spy.assert_not_called()
+
+        res = fn(dev.batch_execute(tapes))
+        assert res.shape == (1, 2)
+
+        expected = [
+            -c * np.cos(x) * np.sin(y) - np.sin(x) * (a + b * np.sin(y)),
+            b * np.cos(x) * np.cos(y) - c * np.cos(y) * np.sin(x),
+        ]
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_trainable_coeffs(self, mocker, tol):
+        """Test trainable Hamiltonian coefficients"""
+        dev = qml.device("default.qubit", wires=2)
+        spy = mocker.spy(qml.gradients, "hamiltonian_grad")
+
+        obs = [qml.PauliZ(0), qml.PauliZ(0) @ qml.PauliX(1), qml.PauliY(0)]
+        coeffs = np.array([0.1, 0.2, 0.3])
+        H = qml.Hamiltonian(coeffs, obs)
+
+        weights = np.array([0.4, 0.5])
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(weights[0], wires=0)
+            qml.RY(weights[1], wires=1)
+            qml.CNOT(wires=[0, 1])
+            qml.expval(H)
+
+        a, b, c = coeffs
+        x, y = weights
+        tape.trainable_params = {0, 1, 2, 4}

```

Comment:
Any reason it is not using all 3 coeffs of H?

Reply:
Oh... very good question 🤔 I remember doing this deliberately, but for the life of me cannot remember the reason!

Reply:
Oh! Sorry I misread this line. At first I thought it was `==`, but now I realize it is an assignment `=`. 

Basically, I want to test the case where not all Hamiltonian coefficients are trainable, just to make the test more complicated.